Add the migration page

kienstra opened the pull request:
<!--- Please summarize this PR in the title above -->

#### Changes
* The markup and CSS is from Rob's Gist 😄  
* Migrates Block Lab blocks to Genesis Custom Blocks
* Adds a 'Migrate' submenu under `/wp-admin > 'Block Lab'`

![migration-full](https://user-images.githubusercontent.com/4063887/87693998-3f833d80-c753-11ea-8d30-fe2d9fbc2a6c.gif)


#### Testing instructions (updated July 27th)
1. Upload and activate the plugin [block-lab.zip](https://github.com/getblocklab/block-lab/files/4983292/block-lab.zip), or do `composer install && npm install && npm run build`
2. If needed, [populate some Block Lab blocks](https://github.com/getblocklab/block-lab/pull/606#issuecomment-646743052)
3. Either install the GCB plugin: [genesis-custom-blocks.zip](https://github.com/getblocklab/block-lab/files/4983329/genesis-custom-blocks.zip), or do steps 5-7
4. Don't activate it yet
5. `cd path/to/wp-content/plugins/`
6. `git clone https://github.com/studiopress/genesis-custom-blocks` # if you haven't done this already
7. `cd genesis-custom-blocks && composer install && npm install && npm run build` # So that it's available to be activated in the final step
8. Go to `/wp-admin` > Block Lab > Migrate
9. Go through the migration flow 
10. Expected: The Block Lab blocks should now appear in Genesis Custom Blocks:
<img width="958" alt="bl-block-here" src="https://user-images.githubusercontent.com/4063887/87060407-1b15e700-c1d0-11ea-9478-44973b5d19e7.png">

11. Expected: Block Lab blocks in posts should look the same other than icons, and should now be Genesis Custom Blocks blocks:

<img width="1179" alt="icon-genesis-here" src="https://user-images.githubusercontent.com/4063887/87060518-3da80000-c1d0-11ea-9ab6-6bd29aa541e2.png">

12. Here's how to [reverse the migration](https://github.com/getblocklab/block-lab/pull/629#issuecomment-655846402)

kienstra committed:
Add a migration submenu under Block Lab

This is still just a skeleton,
it needs to be populated with JS and CSS,
and a container tag like <div id="bl-migration">

kienstra committed:
Add Rob's CSS from the migration gist

A verbatim copy, just some
formatting like tabs changed.

kienstra committed:
Add a migration script, though it's empty now

Also, enqueue it like the CSS file.
Add a js/migration/ directory for the migration components.

kienstra committed:
Merge branch 'add/full-migration' into add/migration-ui

kienstra committed:
Add the beginning of a working JS file for migration

It now outputs a simple sentence,
but the main structure for outputting something
is in place.

kienstra commented:
There's a lot of work left here, but there's now a 'Migrate' submenu page:

<img width="583" alt="migration-ui-here" src="https://user-images.githubusercontent.com/4063887/86170173-56cdf400-bae0-11ea-842b-40aceecc7b5d.png">

kienstra committed:
Fix failed PHPUnit test

This didn't call the tested function,
so it couldn't pass.

kienstra commented:
If it's alright, I think React will be useful here, and I started the React setup for the Migrate page.

It'd be possible to do this in Vanilla JS, like how the CPT page is now.

But I'd be slower developing for that, mixing JavaScript and PHP. And with React, it'll be easier to test.

kienstra committed:
Add the migration intro to the page

Copying the markup from Rob's gist.
Updated a few style names,
but the styles remain.

kienstra committed:
Add the 'Backup your site' step in migration

Most of these components can be
reused in most of the steps.

kienstra committed:
Make the announment post link open in a separate tab

It wouldn't be good if they
left the migration in the middle of the process

kienstra committed:
Add the 'Update Hooks & API step

This has a 'Previous' link.
I might be able to refactor these steps
into a common step,
the first 2 are pretty similar.

kienstra commented:
There's still a lot of work left, but here's what the migration looks like right now. 

Only 2 of the 5 steps are in place so far.

![migration-for-now](https://user-images.githubusercontent.com/4063887/86295119-856ccd00-bbbb-11ea-88e7-707dea648f1b.gif)

lukecarbis commented:
Looking _amazing_, Ryan! 🤩

kienstra commented:
Thanks, Luke! Thanks for your great work on the UI.

kienstra committed:
Add a Get Genesis Pro migration step

todo: make this only display if ->is_pro()

as it's not needed otherwise.

kienstra committed:
Only show 'Get Genesis Pro' if ->is_pro()

There's no need to show this
for other users.

kienstra committed:
Add the 'Migrate Blocks' step, but it still needs the API

This mainly has the presentation,
but it doesn't connect to an API,
as it doesn't exist yet.

kienstra committed:
Add the step 'Activate Genesis Custom Blocks'

Add a nonced URL that works with a plugin
that exists.

kienstra committed:
Add spaces in Intro copy

kienstra committed:
In the 'activate' link, make it deactivate BL before activating GCB

Without this, the activation fails,
as there are 2 definitions of
block_field() and block_value().

kienstra commented:
There's still a lot of work left, but here's what it looks like now:

![basic-bl-migration](https://user-images.githubusercontent.com/4063887/86835080-3ca49080-c061-11ea-975e-be009e41a006.gif)

lukecarbis commented:
Looking good Ryan! I'm sure @RobStino will drop in with a few design tweaks at some stage, too.

Interesting point on onboarding. Do you think we should somehow try to disable it for migrations? IMO, it's not a huge problem, so if it's going to cause a bunch of extra work, I wouldn't worry about it.

kienstra commented:
Ah, good point about the onboarding. I'll look at using a query var to disable the onboarding after the migration.

lukecarbis commented:
Only if it's easy.

On Wed, 8 Jul 2020 at 10:28, Ryan Kienstra <notifications@github.com> wrote:

> Ah, good point about the onboarding. I'll look at if there's a query var I
> could use to disable the onboarding after the migration.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/getblocklab/block-lab/pull/629#issuecomment-655209188>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIL7Q255JLWINXW7JIUSCDR2O4SNANCNFSM4OHIYMMQ>
> .
>

kienstra commented:
OK, thanks.

kienstra committed:
Add REST API endpoints for post_content and CPT migration

To allow the React UI
to call these in that step.

kienstra committed:
Connect the React UI to the endpoints

Now, this actuall migrates.
But it doesn't handle errors yet,
nor does it install GCB.

kienstra commented:
## Does work
* The migration workflow, including migrating the `block_lab` CPT and the blocks in the `post_content`

![migration-with-steps](https://user-images.githubusercontent.com/4063887/86988085-0d734980-c15d-11ea-9d12-d5f5a29b25f0.gif)


## Doesn't work yet

* Automatically downloading GCB:

![migrates-block](https://user-images.githubusercontent.com/4063887/86987333-6cd05a00-c15b-11ea-8b22-e188b7765c2a.png)

kienstra commented:
To test this, please also clone [GCB](https://github.com/studiopress/genesis-custom-blocks) and do `composer install && npm install && npm run build`. This is so that it can be activated in the last step of the UI.

You also might want to do `wp db export` so you can restore to this backup of your DB.

## Reversing the migration

To reverse the migration, ensure Block Lab is active and do:
```sh
wp shell
( new Block_Lab\Admin\Migration\Post_Content( 'genesis-custom-blocks', 'block-lab' ) )->migrate_all();
( new Block_Lab\Admin\Migration\Post_Type( 'genesis_custom_block', 'genesis-custom-blocks', 'genesis_custom_blocks', 'block_lab', 'block-lab', 'block_lab' ) )->migrate_all();

```

kienstra committed:
Fix PHPUnit and Jest tests

Some PHPUnit tests failed in multisite.
Also, fix failures in Jest tests.

kienstra committed:
Change 'Set' to 'Sets'

kienstra committed:
Merge branch 'add/full-migration' into add/migration-ui

kienstra committed:
Move the Submenu class inside migration/

Most of the migration classes
will be in this directory.

kienstra committed:
Move the REST API registration into a new class

There will be more to add,
so this should keep it better organized.

kienstra committed:
Also migrate draft blocks

Use 'post_status' => 'any'
So this includes drafts,
but not auto-drafts or trashed posts.
https://developer.wordpress.org/reference/classes/wp_query/#status-parameters

kienstra committed:
Change from 'license key' to 'subscription key'

That's the phrase that's used
in the other Genesis Pro products.

kienstra committed:
Add the actual hashed discount code

Thanks to Luke for the function to
get the hash.

kienstra committed:
Allow actually saving the subscription key in the UI

Add a REST API endpoint to
save the subscription key,
and some feedback.

kienstra commented:
What do you think about the messages on submitting the Genesis Pro subscription key?

![empty-here](https://user-images.githubusercontent.com/4063887/87108740-999e7300-c228-11ea-9105-afbf2c84de57.gif)

There's no validation of the key, it just checks that it's a non-empty string.

As you know, the validation logic is in Genesis Custom Blocks.

There's also a message if `update_option()` returns `false`:

>There was an error saving the key

kienstra committed:
Prevent onboarding from appearing in GCB on migration

Like Luke mentioned,
there probably isn't a need for onboarding
if people are migrating from BL.
This depends on a PR for GBC.

kienstra committed:
Change the query var from disable_ to disabled_

kienstra committed:
Change 'discount code' to 'coupon code'

That's how my.wpengine.com calls it:
https://my.wpengine.com/signup?plan=genesis-pro

kienstra committed:
Remove WP-CLI testing script

This used a return value from ->migrate_all()
that doesn't exist anymore.
It used the post IDs that were migrated.
But that was for the initital testing stage,
as storing all of the post IDs in that variable
isn't good if there are thousands of posts.

kienstra committed:
Handle error messages for the block migration

If there's an error in block migration
in post content or in a CPT,
display an error.

kienstra committed:
Add a lighter color to a disabled button

Copied from the color of the
'Developer Notice' button.

kienstra committed:
Move migration class init to Admin

The Migration class
was just to bootstrap the other classes.

kienstra committed:
Correct font size to be like to the gist

There was a font-size: 13px rule in common.css
in WP Core,
which made this smaller than in the Gist.

kienstra committed:
Replace some untranslated strings, fix TypeScript errors

Change the type of the onClick()
handler to have generics.

johnstonphilip commented:
Hey @kienstra - I took this for a spin and this works really well and looks great!

I hit a couple snags, though may not be related. 

1. I attempted to have both plugins active at the same time, which is throwing this:

```
Plugin could not be activated because it triggered a fatal error.


Fatal error: Cannot redeclare block_field() (previously declared in /Users/phil.johnston/Documents/wps/wpengine1/app/public/wp-content/plugins/block-lab/php/helpers.php:20) in /Users/phil.johnston/Documents/wps/wpengine1/app/public/wp-content/plugins/genesis-custom-blocks/php/Helpers.php on line 20

```

2. The "Developer Notice button didn't seem to work (nothing happens upon click)

kienstra commented:
Hi @johnstonphilip,
Thanks a lot for testing this, I really appreciate it.

1. Good point, I still need to fix this issue with both plugins being active.

2. Good catch also, https://github.com/getblocklab/block-lab/pull/629/commits/8f74facec192cb782ac82aac568b1f3141baa3a0#diff-1f12e601c690bb1256c9c71f5989a088R39 adds a link for the developer notice.

Thanks, and have a great weekend!

kienstra committed:
Add more test steps to GetGenesisPro tests

Test adding a subscription key,
and assert the messages.

kienstra committed:
Update some DocBlocks and remove extra space

kienstra committed:
Add props isStepActive and isStepComplete

Before, these were created
inside the components.

kienstra commented on js/migration/components/steps/get-genesis-pro.js:
```diff
@@ -0,0 +1,236 @@
+	const urlMigrateWithoutGenPro = 'https://example.com';
+
+	// @todo: Is this a stable URL? Or should it be changed to https://www.studiopress.com/genesis-pro/
+	const urlGetGenesisPro = 'https://my.wpengine.com/signup?plan=genesis-pro';
```
Is this a stable URL? Or should it be changed to https://www.studiopress.com/genesis-pro/ ?

kienstra committed:
Rename get_discount_code() to get_coupon_code()

This is how StudioPress refers to it.

kienstra committed:
Call expect() instead of Monkey\Functions\expect

use function makes this simpler.

kienstra commented on js/migration/components/steps/get-genesis-pro.js:
```diff
@@ -0,0 +1,234 @@
+const GetGenesisPro = ( { goToNext, isStepActive, isStepComplete, stepIndex } ) => {
+	// @todo: replace this.
+	const urlMigrateWithoutGenPro = 'https://example.com';
+	const urlGetGenesisPro = 'https://my.wpengine.com/signup?plan=genesis-pro';
```
Is this a stable URL? Or should it be changed to https://www.studiopress.com/genesis-pro/ ?

lukecarbis replied:
In this case, sending people directly to the signup page (not another marketing page) is preferred, so leave as is.

kienstra committed:
Change getScrollContainer to use import

Instead of a const descrucuturing.
I probably made that mistake.

kienstra commented:
### Request for review

Hi @lukecarbis and @RobStino,
Could you please review this? The [top comment](https://github.com/getblocklab/block-lab/pull/629#issue-439492116) is up to date.

This migration doesn't actually install Genesis Custom Blocks. That's in a ticket that's scheduled for next sprint, though I'll look at doing that soon.

Here are example error messages from post content migration:

<img width="767" alt="migration-errors-here" src="https://user-images.githubusercontent.com/4063887/87210480-45a09680-c2db-11ea-9f04-237daa1c99a5.png">

Though ideally there won't be errors 😄 

There are also `@todo` comments here to add URLs, like the [Developer Notice](https://github.com/getblocklab/block-lab/pull/629/files#diff-1f12e601c690bb1256c9c71f5989a088R20). Once we know those, I'll add them in.

It'll be interesting how long the post content migration takes for you. It usually takes a few seconds for me.

Thanks, Rob and Luke!

lukecarbis commented on js/migration/components/steps/get-genesis-pro.js:
```diff
@@ -0,0 +1,234 @@
+
+		// @ts-ignore global
+		if ( subscriptionKey === blockLabMigration.couponCode ) {
+			setSubmissionMessage( __( "That looks like your coupon code. Please click 'Get Genesis Pro' and enter your coupon code there.", 'block-lab' ) );
```
Very, very, very nice. Brilliant work.

kienstra replied:
Thanks, Luke!

kienstra committed:
Remove the nonce for disabling onboarding

See the accompanying commit in GCB:
https://github.com/studiopress/genesis-custom-blocks/pull/8/commits/ee94edbc592b4039f62c902cc7b8110923c1e964

lukecarbis reviewed (changes requested):
All the code looks fantastic, of course.

I'm not seeing any migration menu in Block Lab though.
https://ufile.io/1n084xcy

kienstra commented:
Hi @lukecarbis,
Sorry the menu's not displaying. That makes it hard to test 😆 

Does this help?

```diff
diff --git a/php/admin/migration/class-submenu.php b/php/admin/migration/class-submenu.php
index d953c8e..4d1b1a7 100644
--- a/php/admin/migration/class-submenu.php
+++ b/php/admin/migration/class-submenu.php
@@ -135,7 +135,7 @@ class Submenu extends Component_Abstract {
         * @return bool Whether the user can view the migration page.
         */
        public function user_can_view_migration_page() {
-               return current_user_can( 'install_plugins' ) && current_user_can( self::MIGRATION_CAPABILITY );
+               return true;
        }
```

...just to narrow down the problem.

lukecarbis commented:
Ah, damn. Sorry man. Had the wrong branch checked out.

kienstra commented:
All good, there's a lot to juggle here with 2 plugins and all of the setup instructions.

lukecarbis reviewed (changes requested):
<img width="1121" alt="Screen Shot 2020-07-13 at 2 22 29 pm" src="https://user-images.githubusercontent.com/1097667/87271570-ec24ab80-c516-11ea-85f2-f5b1fbf3608b.png">
The last paragraph should have no bottom margin.

<img width="1129" alt="Screen Shot 2020-07-13 at 2 34 44 pm" src="https://user-images.githubusercontent.com/1097667/87271584-fba3f480-c516-11ea-9d41-3e6e55448900.png">
Made a bunch of changes to this section. @RobStino could you please review?

<img width="611" alt="Screen Shot 2020-07-13 at 2 39 38 pm" src="https://user-images.githubusercontent.com/1097667/87271598-078fb680-c517-11ea-8f8c-3174c1d0ac13.png">
Is this button supposed to work yet?

<img width="501" alt="Screen Shot 2020-07-13 at 2 40 59 pm" src="https://user-images.githubusercontent.com/1097667/87271603-0e1e2e00-c517-11ea-973f-5cc54e92cd78.png">
Not yet responsive.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/1097667/87271700-63f2d600-c517-11ea-9393-3c875bb934a0.gif)
Animations between steps are very quick / disorienting.

(happened to fast for a screenshot)
The loading bar during migration seemed blue instead of purple.

lukecarbis commented:
Maybe instead of a loading bar we should show a spinner, since it's all happening awfully fast. Maybe it's only happening fast because we're doing it locally?

kienstra commented:
>Is this button supposed to work yet?

Yeah, the 'Submit' button for the subscription key should work. Is there a console error?

kienstra commented:
>Maybe instead of a loading bar we should show a spinner, since it's all happening awfully fast. Maybe it's only happening fast because we're doing it locally?

Yeah, a spinner might be a good idea. Some steps will probably be really fast, especially the CPT migration. In my local, it almost was too fast to see.

Yeah, maybe the speed is from being in a local environment, but some calls will probably be fast normally.

lukecarbis commented:
I just tried saving the key again and it worked. Must have broken when I messed with the DOM for that mockup. My bad, sorry.

kienstra commented:
No problem, it could have been a bug, who knows 😄

RobStino reviewed (commented):
Hey @kienstra 
This is looking really good.
I tested both without and with a Block Lab Pro license key active and got expected behaviour.
The only thing that I spotted that might be weird is that I got this error in the block editor _after_ the migration.
![image](https://user-images.githubusercontent.com/1578148/87378015-403f9680-c5d0-11ea-9ee5-fdd82619bec7.png)
Now I know that the GCB plugin wasn't installed as that part is yet to be sorted, but as the block had been migrated, I was unsure about seeing the `block-lab` slug in that error notice. Not sure how it works technically, or even if this is an issue, but wanted to call it out just in case.

There are a couple of things that need adjusting:
1. We need a better disabled-state for the buttons. They still look active.
2. The animation between steps is janky. I think that is my hacky CSS (height transitions in CSS are notoriously hard!). Is there something you could do with JS to handle transitions?
3. We should capture a task somewhere to update all the external links before launch.

RobStino commented:
For the disabled state button (#3 above), I mean the next-step buttons and their state prior to the checkbox being checked.
Here is some CSS that I think styles it appropriately:
```
color: var(--color-gray-500);
background-color: var(--color-gray-200);
box-shadow: none;
cursor: not-allowed;
```
For the "allowed state", we'll probably have to add:
```
cursor: pointer;

kienstra commented:
Hi @RobStino,
Thanks for testing this and your comments.

>The only thing that I spotted that might be weird is that I got this error in the block editor after the migration.

>...I was unsure about seeing the block-lab slug in that error notice. 

Right, that shouldn't happen. The migration should have changed that from `block-lab/notice` to `genesis-custom-blocks/notice` in the `post_content`.

1. What post type was that block in? A post, page, etc..?

2. Whats's the version of WordPress, and is the Gutenberg plugin active?

3. Did other blocks migrate OK?

nickcernis commented:
Hey, @kienstra. The migration works smoothly for me and looks great. Nice work!

A couple of things I dug up:

### Restoring from browser backups of an unsaved post breaks blocks

The only way I found of reproducing the “your site doesn't include support for the [blockname] block“ from [Rob's comment](https://github.com/getblocklab/block-lab/pull/629#pullrequestreview-447748164) was to restore an old browser save or autosave after doing the migration:

1. Add a custom block to a page and publish the page.
2. Add another custom block to the page, do not publish the page, and wait ~10 seconds for the browser to back up your changes to local storage.
2. Refresh the page. You should see “The backup of this post in your browser is different from the version below.” and a button to “Restore the backup”.
3. In a new tab, complete the block migration.
4. Return to the page (refresh again if you wish) and click “Restore the backup”.

The migrated block and the unsaved block will be replaced with the old cached blocks, because the backup was made with the old Block Lab block names.

This might not be something you need/want to account for, but just flagging it as a possibility in case it comes up in support.

I would expect to see a similar issue with any post autosaves (“There is an autosave of this post that is more recent than the version below. View the autosave.”) but that requires a manual compare-and-restore step where it's more obvious that the block names have changed because you get to see the diff first.

### Keyboard and screen reader accessibility

I did a basic test for keyboard and screen reader accessibility. These may be separate issues to follow up on:

1. Navigating the page at `wp-admin/edit.php?post_type=block_lab&page=block-lab-migration` by keyboard (in Chrome, click to focus the page, then repeatedly press tab) reveals hidden next steps before I have begun the first step. I would not expect to be able to access later steps until I have completed the first one. It could prove confusing for a screen reader user or someone whose only means of access is by keyboard.

    ![tab-through-steps](https://user-images.githubusercontent.com/647669/87402367-bede0900-c5bb-11ea-8139-ff0b2de4ca8d.gif)

2. Step 2: the “ More details here” links could be more descriptive so that screen reader users know what they're clicking on. [Learn more about accessible link text](https://www.washington.edu/accessibility/links/).

3. There's no audible progress alert for screen reader users. When I click “migrate now” in step 3 I would not be able to tell that the step has ever completed if I can't see the screen. In Genesis, John solved this with `wp.a11y.speak()` calls to announce what's going on:
    - [wp-a11y-speak](https://make.wordpress.org/accessibility/handbook/markup/wp-a11y-speak/)
    - [Genesis a11y.speak source](https://github.com/studiopress/genesis/blob/75377bc8f28004454932c5cdf3733862a74029ec/lib/js/admin.js#L359)

4. The “…and step into the FUTURE ꜰᴜᴛᴜʀᴇ FUTURE.” text in step 4 is cute, but VoiceOver read this as, “and step into the future latin letter small capital f, latin letter small capital u, latin letter small capital t, latin letter small capital u, latin letter small capital r, latin letter small capital e future”, which might annoy screen reader users.

   I would either wrap the text in spans and use CSS to adjust their size (which VoiceOver will read as “step into the future future future” which still gives the echo effect) or just change it to “…and step into the future”, even if it loses some of the personality you were going for here.

    It's not so bad in this case but it's something to be aware of as [special character usage can get much worse for screen readers](https://twitter.com/kentcdodds/status/1083073242330361856)! 

5. The final “activate plugin” step redirects to the plugin page, where it's not clear what has happened as a screen reader user. If you can activate the plugin (and announce “Genesis Custom Blocks plugin activated” or similar) but stay on the migration page that may be less disorientating.

kienstra commented:
Hi @nickcernis,
Thanks so much for your detailed testing. Great idea to do accessibility testing, I really appreciate it.

I'll work on applying your points.

kienstra committed:
Apply styling that Rob suggested

Making hte disabled button darker
and using cursor: pointer.

kienstra committed:
Apply Nick's feedback: Use aria-label attributes for link text

Thanks to Nick for his idea of more accessible
'More details' text:
https://www.washington.edu/accessibility/links/

kienstra committed:
Apply Luke's feedback in 'Get Genesis Pro' step

Rearrange some things,
like the button 'Get Genesis Pro'

kienstra commented:
Hi @lukecarbis and @RobStino,
How does this look for the 'Get Genesis Pro' step?

It applies Luke's suggestions from https://github.com/getblocklab/block-lab/pull/629#pullrequestreview-446978121.

![to-g-pro](https://user-images.githubusercontent.com/4063887/87461840-96163d80-c5d4-11ea-8a88-9f67d6274891.png)

kienstra committed:
Try a <Spinner> in place of the <progress>

Open for discussion,
to see what everyone thinks.

kienstra commented:
Also, what do you think about this use of a spinner instead of the `<progress>` bar?

The progress bar was good, but the post content migration ended up taking most of the time. The CPT migration was too fast to see in my local.

![spinner-here](https://user-images.githubusercontent.com/4063887/87461550-37e95a80-c5d4-11ea-8f27-6d0165ee6ea0.gif)

kienstra committed:
Make changes between steps slightly smoother

Uses an arbitrary max-height to transition.
This isn't ideal, but it prevents needed a long JS solution.
https://css-tricks.com/using-css-transitions-auto-dimensions/#technique-1-max-height

kienstra commented:
Changing steps is now a little smoother:

![chaning-steps](https://user-images.githubusercontent.com/4063887/87468034-7f74e400-c5de-11ea-9660-3f45d0e518e6.gif)

This [transitions the max-height](https://css-tricks.com/using-css-transitions-auto-dimensions/#technique-1-max-height).

It's hard to transition from a defined height to `height: auto`.

An alternative would be a [more complex JS solution](https://css-tricks.com/using-css-transitions-auto-dimensions/#technique-3-javascript).

kienstra committed:
Make the Genesis Pro tiles a responsive

At 1000px and lower,
they will stack on top of each other.

kienstra commented:
This step is a bit more responsive now:

![them-tiles](https://user-images.githubusercontent.com/4063887/87470372-48a0cd00-c5e2-11ea-8bf9-899306e9f388.gif)

kienstra committed:
Don't render the content of inactive steps

As Nick pointed out,
it was possible to hit tab and go to
inactive steps.

kienstra committed:
Add 'speak' messages to update on the migration

Thanks to Nick's suggestion,
this gives some idea of progress.

kienstra committed:
Keep 'FUTURE', just wrap it in a span

As Nick suggested,
wrap it in a span and style it with
smaller font-size.

kienstra commented:
>3. We should capture a task somewhere to update all the external links before launch.

Sure, I opened BH-452 for that.

kienstra commented:
This is ready for another review.

RobStino commented:
Hey @kienstra ,
Thanks for all those edits. This looks good to me.
I feel like I might want to tweak a few things around padding and transition durations. I'll work on a separate PR for these though. That would be the best approach hey?

nickcernis commented:
Accessibility is much better now. Nice job!

kienstra commented:
Hi @RobStino,
Thanks a lot for looking at this.

>I feel like I might want to tweak a few things around padding and transition durations. I'll work on a separate PR for these though. That would be the best approach hey?

Sure, that sounds good. A PR to the `add/full-migration` branch would be great.

kienstra commented:
Hi @nickcernis,
Thanks a lot for your detailed review. Sorry for not responding to you about [your suggestions](https://github.com/getblocklab/block-lab/pull/629#issuecomment-658062873).

Great point about restoring from browser backups. Thanks for your steps to reproduce, I also saw that.

Of your 5 suggestions on accessibility, I applied all of them except 5:

>The final “activate plugin” step redirects to the plugin page, where it's not clear what has happened as a screen reader user. If you can activate the plugin (and announce “Genesis Custom Blocks plugin activated” or similar) but stay on the migration page that may be less disorientating.

It's a really good suggestion. But due to a quirk in how Genesis Custom Blocks and Block Lab have some of the same functions, they can't be active at the same time. So as part of activating GBC, it has to deactivate Block Lab.

Thanks for your great suggestions.

mindctrl commented:
"Backup your site" should probably be "Back up your site"... backup is a noun.

kienstra commented:
Ah, good point!

kienstra committed:
Change 'backup' to 'back up'

Also change the name of the component,
to BackUpSite.

kienstra commented:
'backup' is now changed to 'back up':

![back-up-site](https://user-images.githubusercontent.com/4063887/87699513-94768200-c75a-11ea-9d13-1d790e8a7006.png)

lukecarbis reviewed (approved):
@kienstra 

Spinner looks great.

Can we change "Save your Genesis Pro Key / Submit" to "Paste your Genesis Pro Key / Save"

Like @RobStino suggested, I have a few animation / design tweaks in mind, but happy to move them to a separate PR.

lukecarbis commented:
Feel free to merge with or without the suggested copy change.

kienstra committed:
Change 'Save your...' to 'Paste your', and make button 'Save'

Luke had a good idea to update
the copy for this.

kienstra commented:
Hi @lukecarbis,
Thanks, good idea!

It's now changed to 'Paste your...' and 'Save':

![paste-save-here](https://user-images.githubusercontent.com/4063887/87741338-e4c90080-c7a9-11ea-81a0-597144cd59d3.png)

sleepykate commented on js/migration/components/intro.js:
```diff
@@ -0,0 +1,55 @@
+				<p>
+					{ __( 'As part of this move, we have been working on a new plugin that is based on what we developed at Block Lab.', 'block-lab' ) }
+					&nbsp;
+					{ __( 'Genesis Custom Blocks is now the home of all our custom block efforts and what we have planned is very very cool!', 'block-lab' ) }
```
Optional polish: Change "very very" to "very, very"

sleepykate commented:
Optional polish:

- Change "ok" to "okay" throughout
- Change "&" to "and" throughout
- Do a quick pass on headers and CTAs for consistent capitalization (sentence case vs. title case)

kienstra commented:
Thanks, @sleepykate! Good ideas.

kienstra commented:
@RobStino and @lukecarbis,
What do you think about [those suggestions](https://github.com/getblocklab/block-lab/pull/629#issuecomment-665271278)? 

I could apply them today.

Thanks!

lukecarbis commented:
@sleepykate Thanks for these suggestions!

Personally I always prefer `okay` to `ok`, too. However, WordPress tends to go with OK (capitalised) ([examples here](https://translate.wordpress.org/projects/wp/dev/en-au/default/?filters%5Bstatus%5D=either&filters%5Boriginal_id%5D=988&filters%5Btranslation_id%5D=2986019)).

On the other hand… all those `OK` instances are button actions, but ours are in the middle of sentences. Given this, I'd prefer to go with Katie's suggestion and **change them to `okay`**.

Regarding the ampersand, my opinion is that ampersands are good in headers, but not body text. So keep "Update Hooks & API" but change lines where ampersands are used in that way.

kienstra commented:
Thanks, @lukecarbis! I'll push those changes to https://github.com/getblocklab/block-lab/pull/660.
